File: app/src/test/java/com/mariotti/developer/futureclock/controllers/AlarmManagementControllerTest.java
package com.mariotti.developer.futureclock.controllers;

import com.mariotti.developer.futureclock.models.Alarm;
import com.mariotti.developer.futureclock.models.WeekDay;

import org.junit.Before;
import org.junit.Test;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.UUID;

import static org.junit.Assert.assertNotNull;

public class AlarmManagementControllerTest {
    List<Alarm> mAlarms = new ArrayList<>();
    Calendar today;
    SimpleDateFormat mSimpleDateFormat = new SimpleDateFormat("HH:mm dd/MM/yyyy z");

    @Before
    public void init() {
        today = Calendar.getInstance();
        String timezone = today.getTimeZone().getDisplayName();
        Alarm alarm1 = new Alarm(UUID.randomUUID(), 18, 7,
                new int[]{WeekDay.INSTANCE.getMONDAY(), WeekDay.INSTANCE.getFRIDAY()},
                timezone, true);
        Alarm alarm2 = new Alarm(UUID.randomUUID(), 10, 34,
                new int[]{WeekDay.INSTANCE.getSATURDAY(), WeekDay.INSTANCE.getSUNDAY()},
                timezone, true);
        Alarm alarm3 = new Alarm(UUID.randomUUID(), 5, 11,
                new int[]{WeekDay.INSTANCE.getTHURSDAY()},
                timezone, true);
        mAlarms.add(alarm1);
        mAlarms.add(alarm2);
        mAlarms.add(alarm3);
    }

    @Test
    public void getNextAlarmNotNullTest() {
        Alarm alarm = AlarmManagementController.getNextAlarm(mAlarms);
        assertNotNull(alarm);
        System.out.println(alarm.toString());
    }

    @Test
    public void getNearestDayForAlarmNotNullTest() {
        for (Alarm alarm : mAlarms) {
            Calendar day = AlarmManagementController.getNearestDayForAlarm(alarm, today);
            assertNotNull(day);
            System.out.println(mSimpleDateFormat.format(day.getTime()));
        }
    }
}

File: app/src/test/java/com/mariotti/developer/futureclock/controllers/AlarmManagementControllerTest.kt
package com.mariotti.developer.futureclock.controllers

import com.mariotti.developer.futureclock.models.Alarm
import com.mariotti.developer.futureclock.models.WeekDay
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Before
import org.junit.Test
import java.text.SimpleDateFormat
import java.util.*

class AlarmManagementControllerTest {
    lateinit var testAlarms: List<Alarm>
    var today: Calendar = Calendar.getInstance()
    var simpleDateFormat = SimpleDateFormat("HH:mm dd/MM/yyyy z")

    @Before
    fun init() {
        // it is 18:00 31/12/2016 - is a Saturday
        today.set(2016, Calendar.DECEMBER, 31, 18, 0)

        val timezone = today.timeZone.getDisplayName(false, TimeZone.SHORT)
//        val alarm1 = Alarm(UUID.randomUUID(), 18, 7, intArrayOf(WeekDay.MONDAY, WeekDay.FRIDAY),
//                timezone, true)
//        val alarm2 = Alarm(UUID.randomUUID(), 10, 34, intArrayOf(WeekDay.SATURDAY, WeekDay.SUNDAY),
//                timezone, true)
//        val alarm3 = Alarm(UUID.randomUUID(), 5, 11, intArrayOf(WeekDay.THURSDAY),
//                timezone, true)
//        val alarm4 = Alarm(UUID.randomUUID(), 22, 34, intArrayOf(), timezone, false)
//        testAlarms = listOf(alarm1, alarm2, alarm3, alarm4)
    }

    @Test
    fun getNextAlarmTest() {
        val activeTestAlarms = testAlarms.filter { it.active }

        val alarm = getNextAlarm(activeTestAlarms, today)
        val alarmExpected = testAlarms[1] // is alarm2 because of Sunday 10:34
        assertEquals(alarmExpected, alarm)
    }

    @Test
    fun getNextAlarmNotNullTest() {
        val alarm = getNextAlarm(testAlarms)
        assertNotNull(alarm)
    }

    @Test
    fun getNearestDayForAlarmNotNullTest() {
        for (alarm in testAlarms) {
            val day = getNearestDayForAlarm(alarm, today)
            assertNotNull(day)
            println(simpleDateFormat.format(day.time))
        }
    }

    @Test
    fun alarmWithDifferentTimezoneTest() {
        // TODO
    }
}
